src/gui/transferlistwidget.cpp

```diff
@@ -586,39 +586,72 @@ void TransferListWidget::deleteSelectedTorrents(const bool deleteLocalFiles)
     const QVector<BitTorrent::TorrentHandle *> torrents = getSelectedTorrents();
     if (torrents.empty()) return;
 
+    QSet<BitTorrent::InfoHash> hashes;
+    for (const BitTorrent::TorrentHandle *torrent : torrents)
+        hashes << torrent->hash();
+
     if (Preferences::instance()->confirmTorrentDeletion()) {
-        auto *dialog = new DeletionConfirmationDialog(this, torrents.size(), torrents[0]->name(), deleteLocalFiles);
+        auto hashesPtr = std::make_shared<QSet<BitTorrent::InfoHash>>(hashes);
+
+        auto *dialog = new DeletionConfirmationDialog(this, hashes.size(), torrents[0]->name(), deleteLocalFiles);
         dialog->setAttribute(Qt::WA_DeleteOnClose);
-        connect(dialog, &DeletionConfirmationDialog::accepted, this, [dialog, torrents]()
+
+        // avoid deleting torrents that are deleted and then re-added again when
+        // the dialog is open and waiting for user input
+        const QMetaObject::Connection removeHashConnection = connect(
+            BitTorrent::Session::instance(), &BitTorrent::Session::torrentAboutToBeRemoved, dialog
+            , [hashesPtr](const BitTorrent::TorrentHandle *torrent)
         {
-            removeTorrents(torrents, dialog->isDeleteFileSelected());
+            hashesPtr->remove(torrent->hash());
         });
+        connect(dialog, &DeletionConfirmationDialog::accepted, this, [this, dialog, hashesPtr, removeHashConnection]()
+        {
+            disconnect(removeHashConnection);
+            removeTorrents(*hashesPtr, dialog->isDeleteFileSelected());
+        });
+
```

Comment:
Your solution is too complicated, don't you think? I've been thinking about it, too (and I was almost ready to implement it, but you beat me to it). Won't the following one do the job?
```c++
connect(dialog, &DeletionConfirmationDialog::accepted, this, [this, dialog]()
{
    deleteTorrents(getSelectedTorrents(), dialog->isDeleteFileSelected());
});
```

Reply:
I took your suggestion and IMO it will only work correctly when the dialog is modal, so I added an assert for it.

Reply:
Well the assert looks silly, so I turn it into comment instead.